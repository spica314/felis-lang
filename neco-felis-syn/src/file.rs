use crate::{Item, Parse, ParseError, token::Token};

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct File {
    items: Vec<Item>,
}

impl File {
    pub fn items(&self) -> &[Item] {
        &self.items
    }
}

impl Parse for File {
    fn parse(tokens: &[Token], i: &mut usize) -> Result<Option<Self>, ParseError> {
        let mut k = *i;

        let mut items = vec![];
        while let Some(item) = Item::parse(tokens, &mut k)? {
            items.push(item);
        }

        let file = File { items };

        *i = k;
        Ok(Some(file))
    }
}

#[cfg(test)]
mod test {
    use crate::{FileIdGenerator, token::Token};

    use super::*;
    use insta::assert_debug_snapshot;

    #[test]
    fn test_parse_inductive_eq() {
        let mut file_id_generator = FileIdGenerator::new();
        let file_id = file_id_generator.generate_file_id();
        let s = std::fs::read_to_string("../testcases/felis/single/inductive_eq.fe").unwrap();
        let tokens = Token::lex(&s, file_id);

        let mut i = 0;
        let file = File::parse(&tokens, &mut i).unwrap().unwrap();

        assert_debug_snapshot!(file);
        assert_eq!(i, tokens.len());
    }

    #[test]
    fn test_parse_inductive_nat() {
        let mut file_id_generator = FileIdGenerator::new();
        let file_id = file_id_generator.generate_file_id();
        let s = std::fs::read_to_string("../testcases/felis/single/inductive_nat.fe").unwrap();
        let tokens = Token::lex(&s, file_id);

        let mut i = 0;
        let file = File::parse(&tokens, &mut i).unwrap().unwrap();

        assert_debug_snapshot!(file);
        assert_eq!(i, tokens.len());
    }

    #[test]
    fn test_parse_eq_and_nat() {
        let mut file_id_generator = FileIdGenerator::new();
        let file_id = file_id_generator.generate_file_id();
        let s = std::fs::read_to_string("../testcases/felis/single/eq_and_nat.fe").unwrap();
        let tokens = Token::lex(&s, file_id);

        let mut i = 0;
        let file = File::parse(&tokens, &mut i).unwrap().unwrap();

        assert_debug_snapshot!(file);
        assert_eq!(i, tokens.len());
    }
}
