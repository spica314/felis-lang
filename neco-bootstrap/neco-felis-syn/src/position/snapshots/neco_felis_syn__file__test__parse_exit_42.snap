---
source: neco-felis-syn/src/file.rs
assertion_line: 90
expression: file
---
File {
    items: [
        Entrypoint(
            ItemEntrypoint {
                keyword_entrypoint: TokenKeyword {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 1,
                        column: 1,
                    },
                    s: "entrypoint",
                },
                name: TokenVariable {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 1,
                        column: 13,
                    },
                    s: "main",
                },
                semicolon: TokenSemicolon {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 1,
                        column: 17,
                    },
                },
                ext: (),
            },
        ),
        UseBuiltin(
            ItemUseBuiltin {
                keyword_use_builtin: TokenKeyword {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 3,
                        column: 1,
                    },
                    s: "use_builtin",
                },
                builtin_name: TokenString {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 3,
                        column: 14,
                    },
                    s: "syscall",
                },
                keyword_as: TokenKeyword {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 3,
                        column: 24,
                    },
                    s: "as",
                },
                name: TokenVariable {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 3,
                        column: 28,
                    },
                    s: "__syscall",
                },
                semicolon: TokenSemicolon {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 3,
                        column: 37,
                    },
                },
                ext: (),
            },
        ),
        Proc(
            ItemProc {
                ptx_modifier: None,
                keyword_proc: TokenKeyword {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 5,
                        column: 1,
                    },
                    s: "proc",
                },
                name: TokenVariable {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 5,
                        column: 7,
                    },
                    s: "main",
                },
                colon: TokenColon {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 5,
                        column: 12,
                    },
                },
                ty: ArrowNodep(
                    TermArrowNodep {
                        from: Unit(
                            TermUnit {
                                paren_l: TokenParenL {
                                    pos: Pos {
                                        file_id: FileId(
                                            0,
                                        ),
                                        line: 5,
                                        column: 14,
                                    },
                                },
                                paren_r: TokenParenR {
                                    pos: Pos {
                                        file_id: FileId(
                                            0,
                                        ),
                                        line: 5,
                                        column: 15,
                                    },
                                },
                                ext: (),
                            },
                        ),
                        arrow: TokenOperator {
                            pos: Pos {
                                file_id: FileId(
                                    0,
                                ),
                                line: 5,
                                column: 17,
                            },
                            s: "->",
                        },
                        to: Unit(
                            TermUnit {
                                paren_l: TokenParenL {
                                    pos: Pos {
                                        file_id: FileId(
                                            0,
                                        ),
                                        line: 5,
                                        column: 20,
                                    },
                                },
                                paren_r: TokenParenR {
                                    pos: Pos {
                                        file_id: FileId(
                                            0,
                                        ),
                                        line: 5,
                                        column: 21,
                                    },
                                },
                                ext: (),
                            },
                        ),
                        ext: (),
                    },
                ),
                proc_block: ItemProcBlock {
                    brace_l: TokenBraceL {
                        pos: Pos {
                            file_id: FileId(
                                0,
                            ),
                            line: 5,
                            column: 23,
                        },
                    },
                    statements: Then(
                        StatementsThen {
                            head: Expr(
                                Apply(
                                    ProcTermApply {
                                        f: Variable(
                                            ProcTermVariable {
                                                variable: TokenVariable {
                                                    pos: Pos {
                                                        file_id: FileId(
                                                            0,
                                                        ),
                                                        line: 6,
                                                        column: 5,
                                                    },
                                                    s: "__syscall",
                                                },
                                                ext: (),
                                            },
                                        ),
                                        args: [
                                            Number(
                                                ProcTermNumber {
                                                    number: TokenNumber {
                                                        pos: Pos {
                                                            file_id: FileId(
                                                                0,
                                                            ),
                                                            line: 6,
                                                            column: 15,
                                                        },
                                                        s: "231",
                                                    },
                                                    ext: (),
                                                },
                                            ),
                                            Number(
                                                ProcTermNumber {
                                                    number: TokenNumber {
                                                        pos: Pos {
                                                            file_id: FileId(
                                                                0,
                                                            ),
                                                            line: 6,
                                                            column: 19,
                                                        },
                                                        s: "42",
                                                    },
                                                    ext: (),
                                                },
                                            ),
                                            Number(
                                                ProcTermNumber {
                                                    number: TokenNumber {
                                                        pos: Pos {
                                                            file_id: FileId(
                                                                0,
                                                            ),
                                                            line: 6,
                                                            column: 22,
                                                        },
                                                        s: "0",
                                                    },
                                                    ext: (),
                                                },
                                            ),
                                            Number(
                                                ProcTermNumber {
                                                    number: TokenNumber {
                                                        pos: Pos {
                                                            file_id: FileId(
                                                                0,
                                                            ),
                                                            line: 6,
                                                            column: 24,
                                                        },
                                                        s: "0",
                                                    },
                                                    ext: (),
                                                },
                                            ),
                                            Number(
                                                ProcTermNumber {
                                                    number: TokenNumber {
                                                        pos: Pos {
                                                            file_id: FileId(
                                                                0,
                                                            ),
                                                            line: 6,
                                                            column: 26,
                                                        },
                                                        s: "0",
                                                    },
                                                    ext: (),
                                                },
                                            ),
                                            Number(
                                                ProcTermNumber {
                                                    number: TokenNumber {
                                                        pos: Pos {
                                                            file_id: FileId(
                                                                0,
                                                            ),
                                                            line: 6,
                                                            column: 28,
                                                        },
                                                        s: "0",
                                                    },
                                                    ext: (),
                                                },
                                            ),
                                        ],
                                        ext: (),
                                    },
                                ),
                            ),
                            semicolon: TokenSemicolon {
                                pos: Pos {
                                    file_id: FileId(
                                        0,
                                    ),
                                    line: 6,
                                    column: 29,
                                },
                            },
                            tail: Nil,
                            ext: (),
                        },
                    ),
                    brace_r: TokenBraceR {
                        pos: Pos {
                            file_id: FileId(
                                0,
                            ),
                            line: 7,
                            column: 1,
                        },
                    },
                    ext: (),
                },
                ext: (),
            },
        ),
    ],
    ext: (),
}
